core/src/main/java/org/apache/calcite/rel/rules/JoinDeriveIsNotNullFilterRule.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptPredicateList;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.JoinInfo;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.logical.LogicalJoin;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexExecutor;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexSimplify;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.ImmutableIntList;
+import org.apache.calcite.util.Util;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that derives is not null filter from inner join.
+ */
+@Value.Enclosing
+public class JoinDeriveIsNotNullFilterRule
+    extends RelRule<JoinDeriveIsNotNullFilterRule.Config> implements TransformationRule {
+
+  public JoinDeriveIsNotNullFilterRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Join join = call.rel(0);
+    final RelBuilder relBuilder = call.builder();
+    final RelMetadataQuery mq = call.getMetadataQuery();
+    JoinInfo joinInfo = join.analyzeCondition();
+    RelNode newLeftNode = createIsNotNullFilter(join.getLeft(), joinInfo.leftKeys,
+        relBuilder, mq);
+    RelNode newRightNode = createIsNotNullFilter(join.getRight(), joinInfo.rightKeys,
+        relBuilder, mq);
+
+    if (newLeftNode != join.getLeft() || newRightNode != join.getRight()) {
+      List<RelNode> inputs = new ArrayList<>();
+      inputs.add(newLeftNode);
+      inputs.add(newRightNode);
+      RelNode newJoin = join.copy(join.getTraitSet(), inputs);
+      call.transformTo(newJoin);
+    }
+  }
+
+  private RelNode createIsNotNullFilter(RelNode input, ImmutableIntList keys,
```

Comment:
This method would be simpler if it didn't return a RelNode but just operated on the RelBuilder. You wouldn't need to call simplify, or check whether the condition is true. It's quite possible it would be just one line.

Reply:
Simplify is used to avoid creating redundant IS NOT NULL filter. Otherwise, it would create a new IS NOT NULL filter every time the rule applies.

Reply:
If you call {{RelBuilder.filter}} it will simplify, and will also ignore {{true}} filters. Same effect, fewer lines of code.

Reply:
If I understand correctly, what you propose can only handle the following case:
```
LogicalProject(DEPTNO=[$7])
  LogicalJoin(condition=[=($3, $12)], joinType=[inner])
      LogicalFilter(condition=[IS NOT NULL($3)])
        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```
. But it can not handle:
```
LogicalProject(DEPTNO=[$1])
  LogicalJoin(condition=[=($0, $3)], joinType=[inner])
    LogicalProject(ENAME=[$1], DEPTNO=[+($7, 1)])
      LogicalFilter(condition=[IS NOT NULL($1)])
        LogicalTableScan(table=[[CATALOG, SALES, EMPNULLABLES]])
    LogicalTableScan(table=[[CATALOG, SALES, EMPNULLABLES]])
```
.